### ISSUE TITLE
Make Blazor WebAssembly work on multithreaded runtime

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspnetcore#54365]
Currently, Blazor WebAssembly's internals make use of the historical JS/WASM single-threadedness guarantees as a performance optimization. For example:

 * Render batches are communicated to JS via zero-copy shared memory rather than by serialization
 * Events triggered from JS are always processed synchronously by .NET, so there's no need to queue them

At first glance, it appears to work with `<WasmEnableThreads>true</WasmEnableThreads>`, but that's a mirage: it does *not* guarantee correctness in that mode, because the core single-threadedness assumption is violated. For example, JS might raise event 2 before event 1 finished processing, and by the time event 1 is done, there might not even be an event 2 handler any more and then you have an error.

### Why we're only just encountering this now

1. We didn't have multithreading before
2. Even to the extent that we did have an early preview of multithreading, it worked by running .NET on the browser UI thread, so we could define that as the sync context and then all the same assumptions would remain valid within that sync context. But now .NET WebAssembly has just moved to the new "deputy thread" multithreading model (i.e., all .NET code is on a background worker), these assumptions are no longer valid (e.g., because the UI thread's JS can no longer communicate with .NET in a synchronous, blocking manner).

### Guaranteeing correctness

We've already solved all these problems in our other hosting environments, Server and WebView, because they've always been innately asynchronous from the beginning. We would need to generalize/port all these mechanisms so they are shared with or copied by the WebAssembly renderer, JS interop, eventing system, etc.:

- [ ] Change rendering to serialize renderbatches to JS instead of doing shared-memory reads
- [ ] Change the rendering flow so that the renderer's `UpdateDisplayAsync` returns a `Task` that doesn't complete until the JS side sends back an explicit acknowledgement of that renderbatch
- [ ] Review JS interop to make sure any low-level assumptions we've made about synchrony during the internals of message passing are replaced by async assumptions
- [ ] Do the above without changing things for the **single-threaded** WebAssembly flavour, since (1) it would be breaking if for example synchronous UI updates became async, and (2) those perf optimizations are there for a reason, and we don't want to make rendering many times more expensive in cases where you'd be serializing huge strings, for example. It's OK to have significant behavioral and performance changes when people opt into multithreading, but not when they don't.

As a broad strategy, I think we can:

 * Factor out some new `WebRenderer` subclass called something like `AsyncWebRenderer` that holds the common logic around serializing renderbatches and accepting the async renderbatch acknowledgements from JS. This can then be used by all of Server, WebView, and multithreaded-WebAssembly.
 * Use our WebView implementation as the reference for how a complete, async-safe rendering and eventing model should work, because this was more recently implemented than Server and is generally simpler and cleaner. There's not a huge amount of code in it, really. For example, if some extra work is needed to queue event notifications, for example, we should see that clearly from the WebView code.

I suspect it's not necessary to have two different builds of `Microsoft.AspNetCore.Components.WebAssembly`. The total amount of code that should differ between the threaded and not-threaded won't be very much. I expect we can just have two different `Renderer` subclasses and some if/else branches in the JS code.

**Comment by pavelsavara** [comment]
Regarding the non-blocking behavior of JS interop. 
- There are methods `BeginInvokeDotNet` and `EndInvokeJS` which have synchronous signature, but the runtime is hacked to treat them as fire-and-forget async messages to deputy thread. If you rename them hack will break. We don't have public API attribute to express that yet. 
- Making `renderBatch` JSImport to return `Task` and have the implementation return a promise should work just fine.
- There is `mono_wasm_gc_lock` and `mono_wasm_gc_unlock` which you call on UI thread (not deputy thread). This is not ideal because it's making the UI involved in GC stop-the-world.
- I wonder if you can offload the DOM event handlers to thread pool rather than deputy thread.

Regarding blocking `.Wait ` throwing PNSE on "deputy" thread
- Let's try to push thru the problems if possible. We could give up bit later. 
- Runtime would probably bring more scenarios. To throw on similar blocking operations which we didn't cover yet.
- At the same time, I'm actively thinking on how to soften the limitation for deputy thread. No conclusion so far.

In last month or so, we switched the implementation of JS interop dispatch from `JSSynchronizationContext` to emscripten internal queue. Therefore
- we could drop `WebAssemblyDispatcher` and replace it with `RendererSynchronizationContextDispatcher` same as on the server side. 
- we can install `RendererSynchronizationContext` and replace the `JSSynchronizationContext` which is installed by default on main thread.

I'm not clear how you would ship 2 different flavors of the components, if you start using `#ifdef` ?

**Comment by Xyncgas** [comment]
> Change rendering to serialize renderbatches to JS instead of doing shared-memory reads

Are you really going to enable multi-threading at the cost of performance, serialization round-trip can potentially greatly reduce the amount of components displayed on screen before it's unresponsive to users, I saw UI frameworks using blazor having their VirtualDOM and render their VirtualDOM in a for loop recognizing the elements in the VirtualDOM and creating a component for every elements, these framework allows dev to compose and generate a large number of components, potentially every button has multiple components and every cell in a calender has multiple components and that is adding up to a lot components for blazor to render quickly, if blazor is sacrificing a lot performance in large quantity of components being rendered, I worry it would be cutting off an entire line of UI frameworks that compose the UI their way to use .NET through blazor and that's a crowd of audience

**Comment by SteveSandersonMS** [comment]
> Are you really going to enable multi-threading at the cost of performance

No, please see the original issue description. This point was addressed.

**Comment by pavelsavara** [comment]
> Regarding blocking `.Wait ` throwing PNSE on "deputy" thread
>     * Let's try to push thru the problems if possible. We could give up bit later.

This shows how many scenarios would have to be avoided if we don't allow blocking .Wait
https://github.com/dotnet/runtime/pull/98802/files

It seems that's too many.

>     * At the same time, I'm actively thinking on how to soften the limitation for deputy thread. No conclusion so far.

I'm very close to enable blocking `.Wait` on deputy (main) thread when running async code.
https://github.com/dotnet/runtime/pull/99422

Our team conclusion so far is, that we will disable synchronous `[JSExport]` on MT to avoid broad class of deadlocks. 
That would be different PR.

**Comment by pavelsavara** [comment]
it would be good to make following `[JSExport]`s async and return `Task`

```
public static string? InvokeDotNet
public static void UpdateRootComponentsCore
private static void ReceiveByteArrayFromJS
```

So that we could disable support for synchronous JSExport completely.

If we are unable to do that, please be aware that any managed code inside of those calls will throw PNSE on blocking `.Wait`

**Comment by tariqalsoahmed** [comment]
Hi @SteveSandersonMS and team,

Our production Blazor app is working and performing nearly perfect!  I'm sure Blazor WASM multi-threading is difficult due to violating single-threaded assumptions.

As a short-term solution ... Have you considered **starting** Blazor on a single non-blocking background thread?

SSR is working beautifully... only prob is Total Blocking Time (TBT) from Blazor.start() blocking UI thread.

https://eatinglove.com/recipe-boost/1096211/The-Easiest-Beef-Pho

Presumably Blazor could start() on a background thread, then marshal to UI thread when WASM state is completely loaded?  At worse DOM flickers but TBT should be reduced to under a second for normal scenarios.  While Blazor background thread loads, DOM simply behaves as normal HTML page. This might be favorable performance profile for any app optimized for SSR.

https://pagespeed.web.dev/analysis/https-eatinglove-com-recipe-boost-1096211-The-Easiest-Beef-Pho/ct3g2rbt3f?hl=en&form_factor=desktop

![image](https://github.com/dotnet/aspnetcore/assets/50665380/acdf8543-bad0-4217-831b-81ed9e7805a0)

Would love to hear your thoughts!

**Comment by pavelsavara** [comment]
Here is a demo how to start wasm single-threaded runtime on separate thread https://github.com/dotnet/runtime/issues/95452

**Comment by tariqalsoahmed** [comment]
> Here is a demo how to start wasm single-threaded runtime on separate thread [dotnet/runtime#95452](https://github.com/dotnet/runtime/issues/95452)

has anyone demo calling **Blazor.start()** from Web Worker separate thread?  Not sure it's even possible but great performance enhancement.

**Comment by jirisykora83** [comment]
> > Here is a demo how to start wasm single-threaded runtime on separate thread [dotnet/runtime#95452](https://github.com/dotnet/runtime/issues/95452)
> 
> has anyone demo calling **Blazor.start()** from Web Worker separate thread? Not sure it's even possible but great performance enhancement.

I don't think it will work because i believe you cannot access DOM directly from Web Worker. They may try make "proxy" to streamline changes (possibly use same technology as for server-side interactive mode?). But hard to say how fast it would be. And it would be nice, BUT when they actually introduce multi-thread support for blazor this mode becomes obsolved (?). Also, this change would only demo because it will break many nugget packages.

**Comment by mkArtakMSFT** [comment]
Related #58874

**Comment by MattParkerDev** [comment]
> Change rendering to serialize renderbatches to JS instead of doing shared-memory reads

Instead of serializing render batches, could SharedArrayBuffers be used? They are available from the main thread and workers. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer

The caveat is that the window must be in a secure context.

![Image](https://github.com/user-attachments/assets/ca75c3b7-b9b4-4f34-9bc7-b7b36e87da7b)

It may not be feasible to require everyone who wishes to use Blazor multi-threaded to apply these restrictions to their site.

**Comment by LostBeard** [comment]
This, moving the .Net runtime to workers when multithreading is enabled, kills 1 to 1 Javascript interop in [SpawnDev.BlazorJS](https://github.com/LostBeard/SpawnDev.BlazorJS) because being able to handle Javascript events in SpawnDev.BlazorJS requires `DotNet.invokeMethod` and that no longer works when .Net now runs in a worker. Synchronous calls into the .Net runtime allow my Blazor libraries to handle Javascript events that MUST be handled synchronously, but that is not the only reason the are needed. Is there a way a user can enable threads but keep the .Net runtime running in the main thread so SpawnDev.BlazorJS can actually be 1 to 1 Javascript to .Net interop when threading is enabled? 

Moving .Net runtime to workers makes complete sense and I am all for it. But **please make keeping the .Net runtime on the main thread optional** even when using threading. It would help prevent my users from having the issue being reported here: [LostBeard/SpawnDev.BlazorJS.WebWorkers/issues/27](https://github.com/LostBeard/SpawnDev.BlazorJS.WebWorkers/issues/27)

## Important
I have a lot of SpawnDev libraries that rely on SpawnDev.BlazorJS and 1 to 1 Javascript interop. [LostBeard Nugets](https://www.nuget.org/profiles/LostBeard)

[SpawnDev.BlazorJS](https://github.com/LostBeard/SpawnDev.BlazorJS) - 147,122 Nuget downloads

[SpawnDev.BlazorJS.WebWorkers](https://github.com/LostBeard/SpawnDev.BlazorJS.WebWorkers) - 90,448 Nuget downloads Uses SpawnDev.BlazorJS to allow running Blazor in workers including service workers.

[SpawnDev.ILGPU](https://github.com/LostBeard/SpawnDev.ILGPU) - 3,105 Nuget downloads (weeks old and already 124 downloads/day) Uses SpawnDev.BlazorJS for cross platform GPU compute (WebGPU, WebGL, Wasm, Cuda, OpenCL, CPU, P2P).

[SpawnDev.ILGPU.ML](https://github.com/LostBeard/SpawnDev.ILGPU.ML) - (Not yet released) Uses SpawnDev.BlazorJS for cross platform machine learning (WebGPU, WebGL, Wasm, Cuda, OpenCL, CPU, P2P).

## Please 🙏
No matter what, PLEASE do not make running the Blazor Wasm .Net runtime in workers required even when Wasm threading is not enabled or it will KILL ALL of my projects that require 1 to 1 Javascript interop. These are dozens of projects and Nuget packages with 322,466 downloads and counting. 


## Recommendation
Allow keeping the .Net runtime on the main window thread. This would be great.  
`<WasmMainThreadOnWorker>false</WasmMainThreadOnWorker>`

**Comment by pavelsavara** [comment]
> SpawnDev.BlazorJS uses `DotNet.invokeMethod` and that no longer works because .Net now runs in a worker. 

The **only** implementation option to make that happen is spin-wait on the UI thread. 
Very ugly, deadlocks, lions!, not supported, frozen browser rendering, networking, bad!
I can imagine making that opt-in, but I don't love it.

> Synchronous calls into the .Net runtime allow my Blazor libraries to handle Javascript events that MUST be handled synchronously

@LostBeard please be very specific, what scenarios ?

**Comment by nealles** [comment]
I totally agree with LostBeard. I understand why code runs in workers, but I think we should have the ability to decide what runs on the main thread and what runs in workers. Workers have a lot of limitations.

Our application was designed to work on the main thread, and when I added WasmEnableThreads, I didn't realize it would break the application completely because it requires a full redesign. I just wanted the ability to use a couple of threads to run heavy functions in workers — not to run the entire application there.

Also, I'd really like to use LostBeard's library, but I can't right now, which is very frustrating. Is there any way to fix this?

**Comment by LostBeard** [comment]
@pavelsavara  
Javascript events. `preventDefault()` requires a synchronous interop, as do many other events. SpawnDev.BlazorJS supports the entire Javascript Web API and all Javascript data types (even has Union). If you don't know why synchronous calls are required then you don't code much for the browser if at all.

101 Repos https://github.com/LostBeard most of which are Blazor Wasm and most of which require SpawnDev.BlazorJS and synchronous .Net runtime on the main thread. There are near countless examples there.

I wrote a Machine Learning library for Blazor Wasm as a fully featured alternative to Microsoft's Onnx Web Runtime (supports TurboQuant already, and 7 model formats),  [SpawnDev.ILGPU.ML](https://github.com/LostBeard/SpawnDev.ILGPU.ML), simply because Microsoft ignores issues like: [github.com/microsoft/onnxruntime/issues/26107](https://github.com/microsoft/onnxruntime/issues/26107#issuecomment-4084145576)

**Comment by LostBeard** [comment]
You CANNOT do FULL Javascript interop with .Net without synchronous calls between Javascript and .Net.

**Comment by LostBeard** [comment]
My main concern is that this doesn't become required; where the .Net Wasm Runtime CANNOT run in the window/main thread. I can understand the issue with threading and running the .Net runtime on the main thread. Honestly between my SpawnDev.BlazorJS.WebWorkers library that lets Blazor run in workers and SpawnDev.ILGPU that lets me run C# on the WebGPU, WebGL, Wasm... I don't really need threads at all. It is affecting a user of mine and I see a slippery slope that could kill my long term projects. That is my primary concern. 🖖 (FYI, my coding skills significantly outweigh my human to human communication skills.)

**Comment by LostBeard** [comment]
Relevant since you are using threading in Blazor Wasm now... I found a major bug in browser Atomics (found it while writing a fibers dispatcher for Wasm in workers): https://issues.chromium.org/issues/495679735

**Comment by LostBeard** [comment]
## The Deputy Thread Model Must Not Be the Only Path Forward

I want to consolidate my concerns into a single, clear post. I've been building the SpawnDev Blazor WebAssembly ecosystem for years, and the architectural direction being taken here threatens not just my projects but an entire category of applications that the Blazor WASM community depends on.

### Who I Am and Why This Matters

I'm the author of the [SpawnDev](https://github.com/LostBeard) library ecosystem — 41 NuGet packages, 323,000+ total downloads, spanning JS interop, WebWorkers, GPU compute, P2P networking, ML inference, cryptography, media processing, and more. All of it is built on [SpawnDev.BlazorJS](https://github.com/nicholasVilela/SpawnDev.BlazorJS) (147,000+ downloads), which provides full, high-fidelity 1-to-1 C# wrappers for JavaScript APIs.

This ecosystem exists because Microsoft's built-in JS interop is incomplete. SpawnDev.BlazorJS fills the gaps — typed wrappers for WebRTC, WebGPU, WebTorrent, Canvas, Crypto, IndexedDB, Streams, and dozens of other browser APIs. [SpawnDev.BlazorJS.WebWorkers](https://www.nuget.org/packages/SpawnDev.BlazorJS.WebWorkers) (90,000+ downloads) already provides multi-threading for Blazor WASM — without requiring the Deputy Thread model — by keeping the main .NET instance on the UI thread and dispatching work to Web Workers explicitly.

### The Technical Problem

The Deputy Thread model moves the entire .NET runtime to a background Web Worker. This severs synchronous communication between JavaScript and .NET. `DotNet.invokeMethod` — the synchronous JS-to-.NET call — no longer works because the browser's UI thread is forbidden from blocking on a worker thread.

This is not a minor inconvenience. It is a fundamental incompatibility with the JavaScript specification.

### Why "Just Use Async" Is Not a Solution

The JavaScript event model requires synchronous handling in many critical scenarios. These are not edge cases — they are core browser functionality:

- **`event.preventDefault()`** — Must be called synchronously within the event handler. If you `await` a response from a worker thread, the browser has already processed the default action. You cannot prevent a form submission, cancel a drag, or block a navigation asynchronously.

- **`event.stopImmediatePropagation()`** — Same constraint. By the time an async response arrives, other listeners have already fired.

- **`beforeunload`** — The return value must be set synchronously. There is no async path.

- **Synchronous property getters/setters** — Many JS APIs expose synchronous properties. A 1-to-1 C# wrapper must be able to read and return these values synchronously.

- **`onmessage` handlers that return values** — Some messaging protocols expect a synchronous return.

These are not SpawnDev-specific requirements. These are requirements of the web platform itself. Any library that aims to provide complete JavaScript interop — including Microsoft's own — will face this wall.

### What I Am Asking For

I am **not** asking you to abandon the Deputy Thread model. I understand its value for applications that need `System.Threading` primitives without blocking the UI.

I am asking for **a supported configuration that allows .NET to remain on the browser's main thread while still enabling background thread pools**. Something like:

```xml
<PropertyGroup>
    <WasmEnableThreads>true</WasmEnableThreads>
    <WasmKeepMainThreadOnUI>true</WasmKeepMainThreadOnUI>
</PropertyGroup>
```

In this mode:
- The main .NET entry point (`Program.Main`) runs on the UI thread
- Synchronous JS interop works exactly as it does today
- `Task.Run` and thread pool work dispatch to background Web Workers
- Libraries like SpawnDev.BlazorJS continue to function correctly
- Developers who need full Deputy Thread isolation can still opt into it

This is not a novel architecture. It is how SpawnDev.BlazorJS.WebWorkers already works — and has worked reliably for years with 90,000+ downloads.

### The Slippery Slope Concern

My deeper concern is that running .NET on the main thread will eventually be deprecated or removed entirely, even for single-threaded builds. If the Deputy Thread becomes the only supported execution model, every Blazor WASM application that depends on synchronous JS interop — not just mine — will break.

I need clarity from the team: **Is running .NET on the browser's main thread a supported, long-term execution model? Or is it being phased out?**

### The Ecosystem at Stake

This affects more than SpawnDev. Any Blazor WASM project that:

- Handles DOM events requiring synchronous responses
- Wraps synchronous JavaScript APIs
- Uses synchronous callbacks from JS to .NET
- Builds real-time applications (WebRTC, WebSocket handlers, canvas rendering)

...will be impacted. The 101 repositories currently depending on SpawnDev.BlazorJS are a visible slice of this, but the broader community impact extends to anyone writing non-trivial JS interop.

### In Summary

The Deputy Thread model solves a real problem for a specific class of applications. But making it the **only** path to threading — or the **only** execution model, period — eliminates an entire category of high-fidelity, high-performance Blazor WASM applications from being possible.

Please give us a way to keep .NET on the main thread while still having access to background threads. The browser already supports this architecture. The community has already proven it works.

Thank you for your time and consideration.

-- Todd Tanner ([@LostBeard](https://github.com/LostBeard))

**Comment by pavelsavara** [comment]
For the ST build, it seems that we will keep supporting running managed code on the UI thread for foreseeable future.

**Comment by LostBeard** [comment]
@pavelsavara Thank you for the clear response. Your confirmation that the ST build will continue supporting managed code on the UI thread is exactly what I needed to hear — that addresses my primary concern.

To be honest, I don't actually need the MT model. Between [SpawnDev.BlazorJS.WebWorkers](https://github.com/LostBeard/SpawnDev.BlazorJS.WebWorkers) for running full .NET in workers and [SpawnDev.ILGPU](https://github.com/LostBeard/SpawnDev.ILGPU) for GPU compute (WebGPU, WebGL, Wasm, CUDA, OpenCL, CPU), I already have the parallel execution I need — without giving up synchronous interop.

My concern was always the slippery slope — that running .NET on the main thread might eventually be deprecated. Knowing that's not the direction you're heading is all I needed.

Appreciate the time you took to explain the threading design constraints. The 2023 design doc is a good read. (dotnet/runtime#126438)

🖖

**Comment by pavelsavara** [comment]
> My concern was that running .NET on the main thread might eventually be deprecated. 

My visibility into future ends in Net12 and I'm not making any commitments here, just hints. 
It's good to hear about your use-cases and see if there are more than handful of such apps.

There would be many benefits if the application code could stop doing synchronous JS calls and in Blazor, the interactive server is proof that it works. There are many Blazor WASM applications that don't do any direct JS interop in user code. All of them would benefit from not blocking UI thread by running managed code on it. 

Part of the people who ask for MT really only need that managed code doesn't freeze the UI in order to deserialize large JSON.

So, we are thinking about a mode, where ST runs on the worker and interacts with DOM from distance, same way as server does.

**Comment by LostBeard** [comment]
@pavelsavara Thank you for the transparency — I appreciate knowing where the thinking is headed rather than finding out after the fact.

I want to push back on the Blazor Server comparison, though. Blazor Server works without synchronous JS interop because it was *designed* from the ground up with that constraint. It doesn't wrap JavaScript APIs 1-to-1. It doesn't need `preventDefault()` to work synchronously from .NET. It operates through a render tree abstraction that was built for async-only communication.

SpawnDev.BlazorJS and libraries like it exist specifically because that abstraction isn't sufficient for the full browser API surface. WebRTC, WebGPU, Canvas, Crypto, Streams, IndexedDB — these are JavaScript APIs that developers need direct access to, and many of them require synchronous interaction. The reason these libraries have 323,000+ downloads is because there is real demand for what the Blazor abstraction layer doesn't cover.

Saying "Blazor Server proves async-only works" is true for Blazor's rendering model. It is not true for the broader use case of .NET as a first-class citizen in the browser — which is what Blazor WASM makes possible and what makes it special.

The applications that don't do direct JS interop would benefit from running on a worker — I agree completely. But the applications that *do* are the ones pushing Blazor WASM into territory no other framework covers: GPU compute in the browser, P2P networking, real-time media, ML inference. These are the applications that make .NET in the browser worth talking about. Removing the main thread option would cut off the most ambitious use cases to optimize for the most common ones.

If ST-on-worker becomes an option, that's great — more options are always good. But if it becomes the *only* option, the same problem I raised about MT applies to ST too.

I'd ask that the team keep "main thread with synchronous interop" as a supported execution model, even if it's not the default. Let developers choose based on what they're building. 🖖

**Comment by nealles** [comment]
@pavelsavara, I don't understand why you compare Blazor WebAssembly and Blazor Server — they are totally different things. And I have to agree with @LostBeard because of this comparison: you don't understand what you are doing this feature for.

You say that what @LostBeard is saying is only his use case, but you are wrong. It is my use case too, and I see it's the use case of several other programmers who liked @LostBeard's suggestion. And this is only the beginning of the feature. You just shipped the feature, and the criticism is already piling up.

I think you are going in a wrong direction. You should listen to the people who want your feature and who want to make it useful. You want to implement it just so you can say you have the feature, and you don't care whether it's actually useful for people or not.

PS: I should apologize for my english as well, I develop much better than I say but I did not think that I will argue with you because I did not understand that you are able to implement the multithreading this way.

**Comment by zayg21-pixel** [comment]
I have a similar issue too. I want to have rendering on canvas using SkiaSharp with content to render generated in a separate thread. It's super simple to render on canvas in a single thread model, but MT, since workers live in deputy threads, I need to have a complicated setup with hacking worker template, spawning ONE MORE thread to pass offscreen canvas to it. And more hacks to yield it. It was an experiment, because all this is way too bad and inconsistent. Problem with event synchronization on top forces me now to rework whole pipline to have 2 dotnet instances in 2 threads and event passing between them. Yes, it's slower, more memory used, but no choice, necause we don't have an option to spawn runtime instance in UI thread.
